Add automatic daily offers to admin CLI

delventhalz opened the pull request:
Adds two commands to the admin cli:

**Command:** `mktadm renew`

Similar to `seed` (formerly known as `submit`), `renew` takes a url and a data file and submits data to the REST API. Specifically, it will submit any `RENEWABLE` Offers, after first closing any open offers that match. This effectively resets rules like `EXCHANGE_ONCE_PER_ACCOUNT`.

**Usage:**

```
bin/mktadm renew -u localhost:8040 -d sawbuck_app/app_data.yaml
```

**Command:** `mktadm schedule`

Uses `cron` to execute other `mktadm` commands on a regular schedule. Can be run hourly, daily, weekly, or monthly. Resubmitting the same command (with the same parameters), will reschedule it. Can also use the `--remove`flag to unschedule a command.

**Usage**
```
bin/mktadm schedule -H "renew -u localhost:8040 -d sawbuck_app/app_data.yaml"
bin/mktadm schedule -D "renew -u localhost:8040 -d sawbuck_app/app_data.yaml"
bin/mktadm schedule --remove "renew -u localhost:8040 -d sawbuck_app/app_data.yaml"
```

You can check your existing cron jobs using:
```
crontab -l
```

Or remove them with:
```
crontab -r
```

Commit:
Rename submit command to seed, and related data to app_data

This more specific command name is necessary as more commands will be
added which will also submit data. Making the app data more generally
named reflects it's future role of containing updates as well as
seed data.

Signed-off-by: Zac Delventhal <zac@bitwise.io>

Commit:
Refactor api submission into seperate CLI service

Signed-off-by: Zac Delventhal <zac@bitwise.io>

Commit:
Refactor data functions into seperate CLI service

Signed-off-by: Zac Delventhal <zac@bitwise.io>

Commit:
Add admin command: renew

This command looks for renewable offers in app data, closes any offers
that match, and then opens new offers. This will have the effect of
"resetting" rules like EXCHANGE_ONCE and EXCHANGE_ONCE_PER_PARTICIPANT.

Signed-off-by: Zac Delventhal <zac@bitwise.io>

Commit:
Update admin data parameter to path properly

Now accepts absolute paths, or will use a path relative to the
marketplace root directory to create an absolute path. This will
be important for adding mktadm renew to a cron job.

Signed-off-by: Zac Delventhal <zac@bitwise.io>

Commit:
Add mktadm schedule command for adding cron jobs

Schedule can add other mktadm commands to cron locally, executing
them on an hourly, daily, weekly, or monthly schedule. Scheduling
the same command twice will reschedule it to the new specified
interval. Use --remove to remove a command from the schedule.

Signed-off-by: Zac Delventhal <zac@bitwise.io>

boydjohnson commented on admin/marketplace_admin/commands/schedule.py:
```diff
@@ -0,0 +1,94 @@
+
+def do_schedule(opts):
+    command = _get_command(opts)
+    crontab = subprocess.check_output(['crontab', '-l']).decode()
```
You could catch the CalledProcessError that is raised here when `crontab -l` is run by a user that doesn't have a crontab set up, and set the crontab variable to the empty string.

boydjohnson reviewed (changes requested):
schedule command assumes that user has crontabs already set.

boydjohnson reviewed (approved):
The reference finding mini-language is really neat.

peterschwarz commented on admin/marketplace_admin/commands/renew.py:
```diff
@@ -0,0 +1,83 @@
+    LOGGER.info('Submitting data to URL: %s', opts.url)
+    fetch = lambda path: api.get(opts.url, path)
+    submit = lambda p, b, a=None: api.post(opts.url, p, b, a)
+    update = lambda p, b, a: api.patch(opts.url, p, b, a)
```
Not exactly the pythonic way, but interesting. Looks `functools.partial` would solve this

delventhalz replied:
Being pythonic is no longer something I am particularly concerned with. Switched to partial.

Commit:
Replace lambdas with calls to partial

In response to:
https://github.com/hyperledger/sawtooth-marketplace/pull/64#pullrequestreview-84785885

Signed-off-by: Zac Delventhal <zac@bitwise.io>
